scraper/node_data.py

```diff
@@ -0,0 +1,209 @@
+        return self.get_data()[url]
+
+    def get_data(self):
+        """
+        Get the node data. Calls the subclass's private method and caches it.
+
+        :return: The node data, keyed by the node's full url.
+        :rtype: dict
+        """
+        if self._data is None:
+            self._data = self._query_data()
+
+        return self._data
+
+    def get_urls(self):
+        """
+        Get the urls of stories to provide to the spider. Calls the subclass's
+        private method and caches it.
+
+        :return: A list of urls.
+        :rtype: list
+        """
+        if self._urls is None:
+            self._urls = self._query_urls()
+
+        return self._urls
+
+    def _query_data(self):
+        """
+        Implement this in the subclass to return the dict of node data keyed
+        by url.
+        """
+        return {}
+
+    def _query_urls(self):
+        """
+        Implement this in the subclass to return the list of urls to stories.
+        """
+        return []
+
+
+class Reader(AbstractNodeData):
+    """
+    The node data Reader. Opens the files and provides the node data and urls.
+    """
+
+    def _query_data(self):
+        """
+        Retrieve the node data from the nodes.json file.
+        """
+        with codecs.open(self.nodes_filename, 'rb') as fp:
+            all_node_data = json.load(fp, encoding=self.encoding)
+
+        return all_node_data
+
+    def _query_urls(self):
+        """
+        Retrieves the urls from urlst.txt.
+        """
+        with codecs.open(self.urls_filename, 'rb', encoding=self.encoding) as fp:
+            urls = fp.readlines()
+
+        return [x.strip() for x in urls]
+
+
+class Writer(AbstractNodeData):
+    """
+    The node data writer. This queries the database and writes the files.
+    """
+
+    def write_nodes_json(self):
+        """
+        Write the nodes.json file.
+        """
+        with codecs.open(self.nodes_filename, 'wb') as fp:
+            json.dump(self.get_data(), fp, encoding=self.encoding)
+
+    def write_urls_txt_file(self):
+        """
+        Write the urls.txt file.
+        """
+        with codecs.open(self.urls_filename, 'wb', encoding=self.encoding) as fp:
+            fp.write('\n'.join(self.get_urls()))
+
+    def _get_connection(self):
+        """
+        Establish a mysql database connection from given db_settings values
+
+        :return: The database connection object.
+        """
+        kwargs = {}
+        if db_settings.mysql_pw:
+            kwargs['passwd'] = db_settings.mysql_pw
+        if db_settings.mysql_user:
+            kwargs['user'] = db_settings.mysql_user
+        if db_settings.mysql_host:
+            kwargs['host'] = db_settings.mysql_host
+        if db_settings.mysql_db:
+            kwargs['db'] = db_settings.mysql_db
+        if db_settings.mysql_port:
+            kwargs['port'] = db_settings.mysql_port
+
+        return MySQLdb.connect(**kwargs)
+
+    def _query_data(self):
```

Comment:
This is the database query for getting node data and urls.

I'm moderately certain this is in latin-1.